### PR TITLE
outputparser: add `Defined` parser to extract a struct from LLM output

### PR DESCRIPTION
**Pull request by erictse**
This introduces an output parser named 'Defined' as an improvement to the existing output parser 'Structured'.

A developer can define a struct with the proper tags which can generate a description of the desired output format for an LLM. The parser then returns a struct of the given type, populated with the data it extracted. The developer experience parallels what's described in this [Build an Extraction Chain tutorial](https://python.langchain.com/v0.2/docs/tutorials/extraction/).

For best results, use this with the tool calling feature of your LLM.

Fixes https://github.com/tmc/langchaingo/issues/171 and https://github.com/tmc/langchaingo/issues/197

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of the primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `Fixes #123`).
- [x] Describes the source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

**Comment by erictse** [comment]
@tmc I've addressed the linter issues, but the `tagalign` issues remain for my test file. This feature may mean looking into disabling the `tagalign` linter may be more worthwhile. It wants to force reordering the `json:` struct tag key so it appears after `describe:`, reducing readability.

**Comment by erictse** [comment]
@FluffyKebab Hi, do you have any feedback on this PR for me? Haven't heard from anyone yet.

**Comment by erictse** [comment]
@eliben Sorry if I'm barking up the wrong tree. I'm not sure who should be reviewing/accepting PRs. Either way, do you have any feedback on this?

**Comment by eliben** [comment]
@erictse this isn't an area of langchaingo I was working on, so I think you'll have to wait for @tmc to review.

File: outputparser/defined.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+// Defined parses JSON output from an LLM into Go structs. A properly tagged
+// struct can generate TypeScript interfaces to help LLMs generate responses
+// that follow the desired JSON structure.
+type Defined[T any] struct {
+	schema string
+}
+
+func NewDefined[T any](source T) (Defined[T], error) {
```
Comment (tmc) [inline outputparser/defined.go:21]: please comment exported symbols.

File: outputparser/defined.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	numTaggedFields := 0
+	for i := 0; i < numFields; i++ {
+		if _, ok := sourceType.Field(i).Tag.Lookup("describe"); ok {
+			numTaggedFields++
+		}
+	}
+	if numTaggedFields == 0 {
+		return empty, errors.New("requires at least 1 field with the struct tag 'describe'")
```
Comment (tmc) [inline outputparser/defined.go:40]: I feel as though we should still operate with zero tagged fields.

File: outputparser/defined.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		result.Write(data)
+	default:
+		return empty, fmt.Errorf("unable to marshal '%s' field type", sourceType.Kind())
+	}
+	return Defined[T]{result.String()}, nil
+}
+
+var _ schema.OutputParser[any] = Defined[any]{}
+
+func (p Defined[T]) GetFormatInstructions() string {
```
Comment (tmc) [inline outputparser/defined.go:59]: please comment on exported symbols.

File: outputparser/defined_test.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+func TestDefined(t *testing.T) {
+	t.Parallel()
+	type Shape struct {
+		Name     string `json:"shapeName" describe:"shape name"`
+		NumSides int    `json:"numSides" describe:"number of sides"`
+	}
+
+	tests := map[string]struct {
```
Comment (tmc) [inline outputparser/defined_test.go:37]: please include more test cases (separate PR ok) that include wantErr: true cases

**Review by tmc — COMMENTED** [review]
Generally LGTM but have a few comments.

File: outputparser/defined.go (hunk 2 above)
Comment (erictse) [reply outputparser/defined.go]: That seems reasonable. I've updated the code, tests, and doc comments accordingly.

**Comment by erictse** [comment]
@tmc I've addressed your comments, including allowing untagged struct fields.

**Comment by tmc** [comment]
I like this broadly, but can’t help but think this represents a limited subset of what we might encode via https://github.com/invopop/jsonschema — which could be reused to handle tool reprs.

thoughts?

**Comment by erictse** [comment]
Using struct tags to generate JSON Schema looks promising! It should probably be on the roadmap.

However, here are reasons for not dropping this pull request:

* More is better (in the unpredictable world of LLMs). Some models may lack or be weaker at JSON Schema support than others. The developer can decide which structured data extraction approach is best for their situation. 
* Easier to use. JSON Schema is more complex, which can take folks longer to work with. This feature is comparable to Pydantic classes, and for developers coming from Python, the concept would be familiar.
* Feature parity. LangChain supports both Pydantic classes and JSON Schema. See [LangChain doc](https://python.langchain.com/v0.2/docs/how_to/structured_output/).
* Faster/cheaper(?) This is speculative, but JSON Schema is more verbose and complex–if it does happen to be more resource intensive for some models, having another approach is helpful.

**Comment by tmc** [comment]
That's all fair, let's do it!

Can you follow on with an example and some docs?

**Pull request merged**